util/licence-checker.py

```diff
@@ -0,0 +1,357 @@
+    def __init__(self, prefix):
+        self.comment_prefix = str(prefix)
+        self.first_line_prefix = self.comment_prefix
+
+    def search_line(self, licence_first_word):
+        return self.comment_prefix + ' ' + licence_first_word
+
+    def expected_full_line(self, licence_line):
+        return self.comment_prefix + ' ' + licence_line
+
+
+class DifferentFirstLineCommentStyle(LineCommentStyle):
+    """Some files have a different allowable prefix for their first line."""
+    def __init__(self, prefix, first_line_prefix):
+        LineCommentStyle.__init__(self, prefix)
+        self.first_line_prefix = first_line_prefix
+
+
+class BlockCommentStyle(object):
+    """Helpers for block-style comments."""
+    def __init__(self, prefix, suffix):
+        self.comment_prefix = str(prefix)
+        self.comment_suffix = str(suffix)
+        self.first_line_prefix = self.comment_prefix
+
+    def search_line(self, licence_first_word):
+        return self.comment_prefix + ' ' + licence_first_word
+
+    def expected_full_line(self, licence_line):
+        return self.comment_prefix + ' ' + licence_line + ' ' + self.comment_suffix
+
+
+SLASH_SLASH = LineCommentStyle("//")
+HASH = LineCommentStyle("#")
+
+# (Priortised) Mapping of file name suffixes to CommentStyle object.
+# If the suffix of your file does not match one of these, it will not be
+# checked.
+COMMENT_CHARS = [
+    # Hardware Files
+    ([".svh", ".sv", ".sv.tpl"], SLASH_SLASH),  # SystemVerilog
+    # Hardware Build Systems
+    ([".tcl"], HASH),  # tcl
+    ([".core", ".core.tpl"], DifferentFirstLineCommentStyle("#", "CAPI=2")),  # FuseSoC Core Files
+    (["Makefile", ".mk"], HASH),  # Makefiles
+    ([".ys"], HASH),  # Yosys script
+    ([".waiver"], HASH),  # Wavier files
+    ([".vlt"], SLASH_SLASH),  # More wavier files
+    ([".vbl"], HASH),  # Verible configuration
+    ([".el", ".el.tpl"], SLASH_SLASH), # Exclusion List
+    ([".do"], SLASH_SLASH),
+    ([".sdc"], HASH),
+    ([".f"], None),  # Dependency Files
+    # Software Files
+    ([".c", ".c.tpl", ".h", ".h.tpl", ".cc", ".cpp"], SLASH_SLASH),  # C,C++
+    ([".S"], SLASH_SLASH),  # Assembly
+    ([".ld"], BlockCommentStyle("/*", "*/")),  # Linker Scripts
+    ([".rs"], SLASH_SLASH),  # Rust
+    # Software Build Systems
+    (["meson.build", "toolchain.txt", "meson_options.txt"], HASH),  # Meson
+    # General Tooling
+    ([".py"], HASH),  # Python
+    ([".sh"], HASH),  # Shell Scripts
+    (["Dockerfile"], HASH),  # Dockerfiles
+    # Configuration
+    ([".hjson", ".hjson.tpl"], SLASH_SLASH),  # hjson
+    ([".yml", ".yaml"], HASH),  # YAML
+    ([".toml"], HASH),  # TOML
+    (["-requirements.txt"], HASH),  # Apt + Python Requirements
+    (["redirector.conf"], HASH),  # nginx config
+    # Documentation
+    ([".md", ".md.tpl", ".html"], None),  # Markdown and HTML
+    ([".css"], BlockCommentStyle("/*", "*/")),  # CSS
+    ([".scss"], SLASH_SLASH),  # SCSS
+    # Unsupported: Explicitly Skip These
```

Comment:
How about skipping binaries?

Reply:
How do you mean "binaries"?

Reply:
The "correct" answer would probably either shell out to `file`, or use a Python wrapper for it. The more appealing solution would be this nice find: https://stackoverflow.com/a/7392391/2948140

Reply:
Ok, I think this might be a follow-up, as that list only deals with file suffixes.

Files that match none of these suffixes do not have the licence requirements enforced on them. Nor do they warn, which may or may not actually be the real issue here.

Reply:
If that's the policy (and it makes sense), then I'd remove the png/svg/ico line as well.

Reply:
If that's the goal (and I agree with that), I'd remove the "# Unsupported: Explicitly Skip These" category, since that's the default action.